### PR TITLE
All API integration changes, etc

### PR DESCRIPTION
**Pull request by shreyagarwaldev**

### CODE REVIEW

File: src/app/app.browser.module.ts [hunk 1]
```diff
@@ -43,7 +46,8 @@ import { DatePickerComponent } from './date-picker/date-picker.component';
     WorkshopsListComponent,
     WorkshopsComponent,
     ContactUsComponent,
-    WorkshopDetailsComponent 
+    WorkshopDetailsComponent,
+	  PageRedirectComponent
```
Comment (sugandhadev) [inline src/app/app.browser.module.ts]: fix the indentation in all the files

File: src/app/app.node.module.ts [hunk 2]
```diff
@@ -42,7 +46,8 @@ import { DatePickerComponent } from './date-picker/date-picker.component';
     WorkshopsListComponent,
     WorkshopsComponent,
     ContactUsComponent,
-    WorkshopDetailsComponent 
+    WorkshopDetailsComponent,
+	  PageRedirectComponent
```
Comment (sugandhadev) [inline src/app/app.node.module.ts]: indentation

File: src/app/contact-us/contact-us.component.scss [hunk 3]
```diff
@@ -1,5 +1,5 @@
 .contact-us {
-    background-image: url('/assets/img/contact.jpg');
+    background-image: url('https://pixelatedplanetcdn.azureedge.net/img/contact.jpg');
```
Comment (sugandhadev) [inline src/app/contact-us/contact-us.component.scss:2]: we need a fall back here if the image is not served

File: src/app/dropdown-menu/dropdown-menu.component.ts [hunk 4]
```diff
@@ -43,4 +45,9 @@ export class DropdownComponent {
   overlay() {
     this.panelOverlay = true;
   }
+  
+  updateList()
+  {
+	  this.selectionChanged.emit("changed");
```
Comment (sugandhadev) [inline src/app/dropdown-menu/dropdown-menu.component.ts]: how are we using this? can we send bool values instead?

File: src/app/home/home.component.html [hunk 5]
```diff
@@ -5,9 +5,9 @@
             <div class="header-content">
                 <div class="header-content-inner">
                     <div class="main-heading" id="homeHeading">Pixelated Planet</div>
-                    <!--<div class="sub-heading">It is a long established fact that a reader will be distracted by the readable content of a page when
-                        looking at its layout.</div>-->
-                    <a class="btn btn-default btn-lg" href="#workshops">Upcoming Workshops</a>
+                    <!-- <div class="sub-heading">It is a long established fact that a reader will be distracted by the readable content of a page when
```
Comment (sugandhadev) [inline src/app/home/home.component.html:8]: Don't remove this code. I'll change and uncomment

File: src/app/home/home.component.ts [hunk 6]
```diff
@@ -7,9 +11,20 @@ import { Component, OnInit } from '@angular/core';
...
+
+  private globalConstants : GlobalConstantsRepository;
+
+  constructor(private globalConstantsRepository:GlobalConstantsRepository) {
+    this.globalConstants = globalConstantsRepository;
+  }
 
   ngOnInit() {
+		var today = new Date();
+		let startDate = today.getFullYear().toString() + "/" + (today.getMonth()+1).toString() + "/" + today.getDate().toString();
```
Comment (sugandhadev) [inline src/app/home/home.component.ts]: use string literals

File: src/app/page-redirect/page-redirect.component.css [hunk 7]
```diff
@@ -0,0 +1,32 @@
+.filters, 
+.list {
+    position: absolute;
+    left:0;
+    width:100%;
+}
+
+.filters {
+    /*top:1.5rem;*/
```
Comment (sugandhadev) [inline src/app/page-redirect/page-redirect.component.css]: remove the commented code

File: src/app/page-redirect/page-redirect.component.html [hunk 8]
```diff
@@ -0,0 +1,7 @@
+<div class="ws-container">
+	<navi-bar></navi-bar>
+
```
Comment (sugandhadev) [inline src/app/page-redirect/page-redirect.component.html]: remove additional lines

File: src/app/services/workshops/workshopRepository.ts [hunk 9]
```diff
@@ -94,41 +78,94 @@ export interface Iitinerary {
...
-            .map((response: Response) => {
-                if (response.status !== 200) {
-                    return Observable.empty<Response>();
-                }
-                return <IWorkshopListDto[]>response.json();
-            }).catch(function (e) {
-                return Observable.empty<Response>();
-            });
+                        .toPromise()
+                        .then(response => {
```
Comment (sugandhadev) [inline src/app/services/workshops/workshopRepository.ts:85]: lol we are back to this one?

File: src/app/services/workshops/workshopRepository.ts [hunk 10]
```diff
@@ -94,41 +78,94 @@ export interface Iitinerary {
 @Injectable()
 export class WorkshopRepository {
 
-    constructor(private http: Http) {
+    public globalConstants;
+
+    constructor(private http: Http, private globalConstantsRepository : GlobalConstantsRepository) {
+        this.globalConstants = globalConstantsRepository;
+        this.getLocations();
+        this.getPhotographers();
```
Comment (sugandhadev) [inline src/app/services/workshops/workshopRepository.ts]: we don't need this

File: src/app/services/workshops/workshopRepository.ts [hunk 11]
```diff
@@ -94,41 +78,94 @@ export interface Iitinerary {
...
     }
 
-    getWorkshopDetails(workshopId: string) {
-        return this.http.get("/assets/ws-details.json")
-            .toPromise()
-            .then(res => <IWorkshopDetails>res.json().data)
-            .then(data => { return data; });
+    getLocations() : Promise<ILocation[]> {
+        let loc = this.globalConstants.getLocations();
+        if(typeof loc == "undefined") {
```
Comment (sugandhadev) [inline src/app/services/workshops/workshopRepository.ts]: if (!loc)

File: src/app/services/workshops/workshopRepository.ts [hunk 12]
```diff
@@ -94,41 +78,94 @@ export interface Iitinerary {
...
+        return this.http.get(this.globalConstants.getWorkshopTypesUrl())
+                        .toPromise()
+                        .then(response => {
+                            return response.json() as string[];
+                        });
+    }
+
+    getWorkshopTypes() : Promise<ILocation[]> {
+        let wTypes = this.globalConstants.getWorkshopTypes();
+        if(typeof wTypes == "undefined") {
```
Comment (sugandhadev) [inline src/app/services/workshops/workshopRepository.ts]: if(!wtypes)

File: src/app/services/workshops/workshopRepository.ts [hunk 13]
```diff
@@ -94,41 +78,94 @@ export interface Iitinerary {
...
+            wTypes.then(workshopTypes =>
+            this.globalConstants.setWorkshopTypes(workshopTypes)
+            );
+        }
+
+        return wTypes;
+    }
+    
+	
+    private getPhotographersInternal() : Promise<IPhotographer[]> {
```
Comment (sugandhadev) [inline src/app/services/workshops/workshopRepository.ts]: not required

File: src/app/services/workshops/workshopRepository.ts [hunk 14]
```diff
@@ -94,41 +78,94 @@ export interface Iitinerary {
...
-                return <ILocation[]>response.json();
-            }).catch(function (e) {
-                return Observable.empty<Response>();
-            });
+    getPhotographers() : Promise<IPhotographer[]> {
+        let p = this.globalConstants.getPhotographers();
+        // if(typeof p == "undefined") {
+        //     p = this.getPhotographersInternal();
+        //     this.globalConstants.setPhotographers(p);
+        // }
```
Comment (sugandhadev) [inline src/app/services/workshops/workshopRepository.ts]: remove commented code

File: src/app/workshop-filter/workshop-filter.component.html [hunk 15]
```diff
@@ -1,13 +1,12 @@
...
-        <dropdown [dataModel]="categories" [buttonLabel]="categoryDropdownLabel" [optionName]="'categories'"></dropdown>
+        <dropdown (selectionChanged)="updateLocationList($event)" [dataModel]="cities" [buttonLabel]="cityDropdownLabel" [optionName]="'location'"></dropdown>
+        <dropdown (selectionChanged)="updateCategoryList($event)" [dataModel]="categories" [buttonLabel]="categoryDropdownLabel" [optionName]="'categories'"></dropdown>
     </div>
 
-
     <div class="price-filter">
-        <input type="number" min="0" step="100" placeholder="Min. price"/>
-        <input type="number" min="0" step="100" placeholder="Max. price"/>
+        <input #minPriceInput type="number" min="0" step="100" placeholder="Min. price" (blur)="updateMinPrice(minPriceInput.value)" (oninput)="updateMinPrice(minPriceInput.value)" (onchange)="updateMinPrice(minPriceInput.value)" (change)="updateMinPrice(minPriceInput.value)"/>
```
Comment (sugandhadev) [inline src/app/workshop-filter/workshop-filter.component.html:8]: do we need both change and onchange?

File: src/app/workshop-filter/workshop-filter.component.ts [hunk 16]
```diff
@@ -55,20 +58,62 @@ export class WorkshopFilterComponent {
     this.minFromDate = new Date();
   }
 
-  createFilterMaps() {
+  updateLocations()
+  {
+    this.cities = [];
+    this.workshopRepository.getLocations().then(locations => { 
+      if(typeof locations != "undefined" && locations != null) {
```
Comment (sugandhadev) [inline src/app/workshop-filter/workshop-filter.component.ts]: just do if(!locations)

File: src/app/workshop-filter/workshop-filter.component.ts [hunk 17]
```diff
@@ -55,20 +58,62 @@ export class WorkshopFilterComponent {
...
+        for (var i = 0; i < locations.length; i++) {
```
Comment (sugandhadev) [inline src/app/workshop-filter/workshop-filter.component.ts]: use foreach

File: src/app/workshop-filter/workshop-filter.component.ts [hunk 18]
```diff
@@ -55,20 +58,62 @@ export class WorkshopFilterComponent {
...
+          var x = <ILocation>locations[i];
+          let labelLoc:string = "";
+          if(x.city != "Multiple" && x.city != "N/A")
```
Comment (sugandhadev) [inline src/app/workshop-filter/workshop-filter.component.ts]: always use !== or === for javascript code. It checks the object type too

Comment (sugandhadev) [reply src/app/workshop-filter/workshop-filter.component.ts]: will lowercase affect his?

File: src/app/workshop-filter/workshop-filter.component.ts [hunk 19]
```diff
@@ -55,20 +58,62 @@ export class WorkshopFilterComponent {
...
+            labelLoc+=x.city + ", ";
+          if(x.state != "Multiple" && x.state != "N/A")
+            labelLoc+=x.state + ", ";
```
Comment (sugandhadev) [inline src/app/workshop-filter/workshop-filter.component.ts]: string literals everywhere

File: src/app/workshop-filter/workshop-filter.component.ts [hunk 20]
```diff
@@ -55,20 +58,62 @@ export class WorkshopFilterComponent {
...
+        }
+      }
+    });
+  }
 
+  updateCategories()
+  {
+    this.categories = [];
+    this.workshopRepository.getWorkshopTypes().then(wTypes => { 
+      if(typeof wTypes != "undefined" && wTypes != null) {
```
Comment (sugandhadev) [inline src/app/workshop-filter/workshop-filter.component.ts]: same as above

File: src/app/workshop-filter/workshop-filter.component.ts [hunk 21]
```diff
@@ -55,20 +58,62 @@ export class WorkshopFilterComponent {
...
+          this.categories.push({label:wTypes[i], value:wTypes[i]});
+        }
+      }
+    });
   }
 
   getFromDate(value: Date) {
     this.fromDate = value;
+	this.angulartics2.eventTrack.next({ action: 'FromDateFilterEvent', properties: { category: 'WorkshopFilterComponent' }});
+
```
Comment (sugandhadev) [inline src/app/workshop-filter/workshop-filter.component.ts]: remove additional line

File: src/app/workshop-filter/workshop-filter.component.ts [hunk 22]
```diff
@@ -81,40 +126,40 @@ export class WorkshopFilterComponent {
...
+  updateCategoryList(value: any)
+  {
+	  this.angulartics2.eventTrack.next({ action: 'CategoryFilterEvent', properties: { category: 'WorkshopFilterComponent' }});
+	
+	  let categories = this.getSelectedFilters('categories');
+	  let workshopTypesList = "";
+	  let first = true;
+	  for(let category of categories)
+	  {
+		  if(first == false)
```
Comment (sugandhadev) [inline src/app/workshop-filter/workshop-filter.component.ts]: if(!first)

File: src/app/workshops-list/workshops-list.component.html [hunk 23]
```diff
@@ -2,23 +2,21 @@
...
                 <div class="location">
                     <i class="fa fa-map-marker" aria-hidden="true"></i>
-                    <span>  {{data.location?.city}},</span>
-                    <span> {{data.location?.state}},</span> 
-                    <span> {{data.location?.country}}</span> 
+                    <span>{{this.formatLocation(data.locationId)}}</span> 
                 </div>
                 <div class="date">
-                    <span>{{data.startDateFirst}}</span>
+                    <span>{{this.formatDate(data.startDate)}}</span>
```
Comment (sugandhadev) [inline src/app/workshops-list/workshops-list.component.html]: we should never call functions from the html. move the logic to typescript files

File: src/app/workshops-list/workshops-list.component.ts [hunk 24]
```diff
@@ -40,36 +36,29 @@ export class WorkshopsListComponent {
...
-                        this.locationList.set(loc.locationId, loc);
-                    })
-                }
-            });
+    formatLocation(locationId){
+        let locationObject = this.workshopRepository.globalConstants.getLocationByLocationId(locationId);
+        if(locationObject){
+            let location:string = "";
+            if(locationObject.city != "Multiple" && locationObject.city != "N/A")
+                location += locationObject.city + ", ";
```
Comment (sugandhadev) [inline src/app/workshops-list/workshops-list.component.ts]: string literals

File: src/app/services/shared/globalConstantsRepository.ts [hunk 25]
```diff
@@ -0,0 +1,98 @@
...
+    
+    public getWorkshopTypes()
+    {
+        return this.workshopTypes;
+    }
+
+    public getPhotographers()
+    {
+        return this.photographers;
+    }
```
Comment (sugandhadev) [inline src/app/services/shared/globalConstantsRepository.ts]: we don't need this

File: src/app/services/shared/globalConstantsRepository.ts [hunk 26]
```diff
@@ -0,0 +1,98 @@
...
+        return this.locationsUrl;
+    }
+
+    public getWorkshopTypesUrl()
+    {
+        return this.workshopTypesUrl;
+    }
+
+    public getPhotographersUrl()
+    {
```
Comment (sugandhadev) [inline src/app/services/shared/globalConstantsRepository.ts]: remove

File: src/app/services/shared/globalConstantsRepository.ts [hunk 27]
```diff
@@ -0,0 +1,98 @@
...
+
+    public getPhotographers()
+    {
+        return this.photographers;
+    }
+
+    public setLocations(locations:ILocation[])
+    {
+        this.locationMap = new Map();
+        for (var i = 0; i < locations.length; i++) {
```
Comment (sugandhadev) [inline src/app/services/shared/globalConstantsRepository.ts:60]: foreach

File: src/app/services/workshops/workshopRepository.ts [hunk 28]
```diff
@@ -24,59 +15,52 @@ export interface ILocation {
...
+	workshopId:string,
+    workshopType:string,
+	numberOfWorkshops:number,
+	startDateFirst:Date,
+	endDateFirst:Date,
+	minCost:number,
+	maxCost:number,
+	costCurrency:string,
+	name:string,
+	description:string,
```
Comment (sugandhadev) [inline src/app/services/workshops/workshopRepository.ts]: not required

File: src/app/services/workshops/workshopRepository.ts [hunk 29]
```diff
@@ -24,59 +15,52 @@ export interface ILocation {
...
+	imageLink:string,
+	locationId:number,
+	primaryPhotographerUserId:number,
```
Comment (sugandhadev) [inline src/app/services/workshops/workshopRepository.ts]: don't need photographers

File: src/app/workshop-filter/workshop-filter.component.html (hunk 15 above)
Comment (sugandhadev) [reply src/app/workshop-filter/workshop-filter.component.html:8]: bug

File: src/app/workshops-list/workshops-list.component.html [hunk 30]
```diff
@@ -2,23 +2,21 @@
...
                     <span>-</span>
-                    <span> {{data.endDateFirst}} </span>
+                    <span> {{this.formatDate(data.endDate)}} </span>
                 </div>
-                <div class="price">$10,000</div>
+                <div class="price">{{data.costCurrency}} {{data.cost}}</div>
```
Comment (sugandhadev) [inline src/app/workshops-list/workshops-list.component.html]: string literal

File: src/app/services/shared/globalConstantsRepository.ts (hunk 27 above)
Comment (shreyagarwaldev) [reply src/app/services/shared/globalConstantsRepository.ts:60]: will do in next CR

**Review by sugandhadev — APPROVED** [review]

**Pull request merged**